Profiles: simplify profile stack trace representation (ASCII fix, conflicts resolved)

christos68k opened the pull request:
### Summary

This PR is essentially @felixge 's #645 with the ASCII diagram updated and rebased on top of current main, resolving conflicts.

I opened it in light of @tigrannajaryan 's comment [here](https://github.com/open-telemetry/opentelemetry-proto/pull/645#issuecomment-3224910458) to try and get it merged before `1.8.0` is out as @felixge will probably not have the time to update #645 before the end of this week.

Please refer to #645 for extensive comments and review history.

aalexand commented on opentelemetry/proto/profiles/v1development/profiles.proto:
```diff
@@ -70,12 +70,18 @@ option go_package = "go.opentelemetry.io/proto/otlp/profiles/v1development";
+//   │                     1-n      │      │
+//   │ 1-n         ┌────────────────┘      │
+//   ▽             │                       │
+// ┌──────────────────┐   n-1   ┌─────────────┐
```
Saying `n-1` for Location -> Mapping and `1-1` for Sample -> Stack seems inconsistent?

It's true that multiple locations can point to the same mapping, but it's also true that multiple samples can point to the same stack?

christos68k replied:
I switched it to `n-1` but there might be other inconsistencies (e.g. Line -> Function) that I didn't touch and that we should open a separate PR for.

christos68k commented:
@tigrannajaryan In light of @felixge being away this week, can we merge this PR instead of #645 (and I'm assuming we can still make it in `1.8.0`) ?

If squash-merging please use 19aa542206ccd85d4aaf6ff0dfaa99be583c7c60 for the commit message and verify that @felixge is credited as the author (I fixed merge conflicts and updated the ASCII diagram but the crux of the work belongs to @felixge).

felixge commented:
Thanks for handling this @christos68k. And sorry I wasn't available. GopherCon is done now, so I'll be responsive again.
